wgpu-hal/src/metal/mod.rs

```diff
@@ -785,6 +785,10 @@ unsafe impl Sync for BindGroup {}
 pub struct ShaderModule {
     naga: crate::NagaShader,
     bounds_checks: wgt::ShaderRuntimeChecks,
+    pub library: Option<metal::Library>,
+    pub function: Option<metal::Function>,
+    pub entry_point: Option<String>,
+    pub num_workgroups: Option<(u32, u32, u32)>,
```

Comment:
If you add a nested enum `ShaderModuleSource` or so, which has `Naga` and `Passthrough` variants, a decent amount of this code will simplify I think.